File: 341. Flatten Nested List Iterator.java
// Time Complexity : O(No. of Nested List)
// Space Complexity : O(No. of levels of Nesting)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

//We will add the whole nestedinteger in stack and then iterate over each, if it is nested iterator then will add to to stack and then iterate over the peek of stack and then add elements to flatten the nested list

public class NestedIterator implements Iterator<Integer> {
    Stack<Iterator<NestedInteger>> s;
    NestedInteger nextEl;
    
    public NestedIterator(List<NestedInteger> nestedList) {
        s = new Stack<>();
        s.push(nestedList.iterator());
    }

    @Override
    public Integer next() {
        return nextEl.getInteger();
    }

    @Override
    public boolean hasNext() {
        while(!s.isEmpty()){
            if(!s.peek().hasNext()){
                s.pop();
            }
            else if((nextEl = s.peek().next()).isInteger()){  
                return true;
            }
            else{
                s.push(nextEl.getList().iterator());
            }
        }
        return false;
    }
}

File: 341. Flatten Nested List Iterator.py
# """
# This is the interface that allows for creating nested lists.
# You should not implement it, or speculate about its implementation
# """
#class NestedInteger:
#    def isInteger(self) -> bool:
#        """
#        @return True if this NestedInteger holds a single integer, rather than a nested list.
#        """
#
#    def getInteger(self) -> int:
#        """
#        @return the single integer that this NestedInteger holds, if it holds a single integer
#        Return None if this NestedInteger holds a nested list
#        """
#
#    def getList(self) -> [NestedInteger]:
#        """
#        @return the nested list that this NestedInteger holds, if it holds a nested list
#        Return None if this NestedInteger holds a single integer
#        """

class NestedIterator:
    def __init__(self, nestedList):
        self.result = []
        self.i = 0
        self.dfs(nestedList)

    def dfs(self, nestedList):
        # Helper function to perform depth-first search on the nested list
        for el in nestedList:
            if el.isInteger():
                self.result.append(el.getInteger())
            else:
                self.dfs(el.getList())

    def next(self):
        # Return the next integer from the result list
        re = self.result[self.i]
        self.i += 1
        return re

    def hasNext(self):
        # Check if there are more integers to be returned from the result list
        return self.i < len(self.result)   
         

# Your NestedIterator object will be instantiated and called as such:
# i, v = NestedIterator(nestedList), []
# while i.hasNext(): v.append(i.next())